### PR TITLE
Consistent resource naming and avoiding of conflicts

### LINKED ISSUE #92: Multiple paths to the same resource
Discovery documents contain a hierarchy of resources. Two resources with the same name can be in multiple places of the tree: e.g. directly under `projects` and then again under `projects/location`. We pick a ~random one currently and build up the names of resources using parent names (except project/location/etc). Is that a valid strategy? Are those endpoints always equivalent? Any preference which one to use?

### PR DESCRIPTION
**Pull request by mikhailshilkov**
One of the fundamental tasks of the provider's codegen is to map DD/API's paths to Pulumi resources. We may use three pieces of information for this mapping:

- Discovery document's resource names
- API paths
- Discovery document schema name

In the current master version, we do the following:

1. Follow the tree of DD resources while walking in depth.
2. Exclude well-known resources like `projects` and `locations`.
3. Concat all the remaining resources that form the parent-child path to the current resource to build the resource name.

This sometimes leads to very long resource names like `OrganizationEnvironmentAnalyticExport`.

Still, there may be multiple paths that map to the same resource name. For example, it may contain `projects` or `projects` and `locations` and we'll get the same name. We bluntly ignore these conflicts and take the ~random path. This is highly undesirable.

This PR introduces the new algorithm:

1. Always take the last resource name only. This ensures that we stay as close as possible to DD naming.
2. Prefer longer paths to shorter path, where a longer path includes all the parameters of the shorter path, but adds some extra parameters. Example: use a path for `/projects/locations` vs. `projects`. We may consider "merging" these paths into a single resource and choosing the path based on user input in the future. This change seems to be non-breaking, so I haven't spent time on it.
3. Otherwise, fail if a conflict occurs (when we try to add a resource but this name is already used by another resource in the same module).
4. Provide manual overrides for well-known cases of (3). I ended up with ~30 overrides across all APIs. The overrides mostly follow the old naming pattern where parent names are prepended to the resource name. It turns out that the majority of overrides are in somewhat more obscure modules that I have to domain expertise in.

As always, feedback is appreciated.

Related to #92 but let's leave it open

### CODE REVIEW

File: examples/simple/index.ts [hunk 1]
```diff
@@ -12,27 +12,16 @@ const cluster = new google.container.v1.Cluster("cluster", {
...
-            "https://www.googleapis.com/auth/service.management.readonly",
-            "https://www.googleapis.com/auth/servicecontrol",
-            "https://www.googleapis.com/auth/trace.append",
-        ],            
-    },
 });
 
 const nodePoolName = "extra-node-pool";
-const pool = new google.container.v1.ClusterNodePool(nodePoolName, {
+const pool = new google.container.v1.NodePool(nodePoolName, {
```
Comment (mikhailshilkov) [inline examples/simple/index.ts:24]: This is the only change in our examples. Main modules are relatively unaffected.

File: provider/pkg/gen/schema.go [hunk 2]
```diff
@@ -624,6 +624,7 @@ func isDeprecated(description string) bool {
 	lowerDesc := strings.ToLower(description)
 	return strings.HasPrefix(lowerDesc, "deprecated") ||
 		strings.Contains(lowerDesc, ". deprecated") ||
+		strings.Contains(description, ". @Deprecated") ||
```
Comment (mikhailshilkov) [inline provider/pkg/gen/schema.go:627]: This isn't strictly related but it resolved a conflict somewhere so I decided to add it.

File: sdk/nodejs/appengine/v1/index.ts [hunk 3]
```diff
@@ -6,32 +6,28 @@ import * as utilities from "../../utilities";
...
     construct: (name: string, type: string, urn: string): pulumi.Resource => {
         switch (type) {
             case "google-native:appengine/v1:App":
                 return new App(name, <any>undefined, { urn })
-            case "google-native:appengine/v1:AppAuthorizedCertificate":
-                return new AppAuthorizedCertificate(name, <any>undefined, { urn })
-            case "google-native:appengine/v1:AppDomainMapping":
-                return new AppDomainMapping(name, <any>undefined, { urn })
-            case "google-native:appengine/v1:AppFirewallIngressRule":
-                return new AppFirewallIngressRule(name, <any>undefined, { urn })
```
Comment (lblackstone) [inline sdk/nodejs/appengine/v1/index.ts:32]: Is it expected that this resource was deleted?

I noticed a few other similar removals as well.

File: sdk/nodejs/firestore/v1beta1/index_.ts [hunk 4]
```diff
@@ -8,31 +8,31 @@ import * as utilities from "../../utilities";
 /**
  * Creates the specified index. A newly created index's initial state is `CREATING`. On completion of the returned google.longrunning.Operation, the state will be `READY`. If the index already exists, the call will return an `ALREADY_EXISTS` status. During creation, the process could result in an error, in which case the index will move to the `ERROR` state. The process can be recovered by fixing the data that caused the error, removing the index with delete, then re-creating the index with create. Indexes with a single field cannot be created.
  */
-export class DatabaseIndex extends pulumi.CustomResource {
+export class Index extends pulumi.CustomResource {
```
Comment (lblackstone) [inline sdk/nodejs/firestore/v1beta1/index_.ts:11]: It looks like this file was renamed to `indext_.ts`. Is that right?

File: provider/pkg/gen/discovery.go [hunk 5]
```diff
@@ -120,14 +139,94 @@ func findResourcesImpl(docName, parentName string, rest map[string]discovery.Res
...
+
+	diff := diffSet(set, other)
+	if len(diff) == 1 && diff.Has("regionId") && set.Has("location") {
+		return true
+	}
+
+	return false
+}
+
+func containsSet(set codegen.StringSet, subset codegen.StringSet) bool {
```
Comment (lblackstone) [inline provider/pkg/gen/discovery.go:191]: Would it make sense to add the contains and diff functions as methods to the StringSet?

**Review by lblackstone — COMMENTED** [review]
Overall, this looks like a great improvement. Just a few questions about edge cases.

File: sdk/nodejs/firestore/v1beta1/index_.ts (hunk 4 above)
Comment (viveklak) [reply sdk/nodejs/firestore/v1beta1/index_.ts:11]: > It looks like this file was renamed to `indext_.ts`. Is that right?

This is expected due to https://github.com/pulumi/pulumi/blob/master/pkg/codegen/nodejs/gen.go#L200.

**Review by viveklak — APPROVED** [review]

File: sdk/nodejs/appengine/v1/index.ts (hunk 3 above)
Comment (mikhailshilkov) [reply sdk/nodejs/appengine/v1/index.ts:32]: Great catch! This wasn't intended - I forgot to update `resourceNamePropertyOverrides`. Doing that now...

File: provider/pkg/gen/discovery.go (hunk 5 above)
Comment (mikhailshilkov) [reply provider/pkg/gen/discovery.go:191]: https://github.com/pulumi/pulumi/pull/7256

**Pull request merged**